### PR TITLE
QuadPlane: Precision Landing(v2)

### PR DESCRIPTION
**Pull request by lukedempsey**
This is a carry-on from @jonathan84clark's original PR (#10378) - adding the precision landing functionality to quadplanes.

I have got it landing consistently in SITL - I'm yet to do a hardware test.

I haven't squashed commits etc yet, as I still have a couple of questions about the code, I'll add them as comments below.

### CODE REVIEW

File: ArduPlane/quadplane.cpp [hunk 1]
```diff
@@ -2577,8 +2661,20 @@ bool QuadPlane::verify_vtol_land(void)
...
+    const bool navigating = pos_control->is_active_xy();
+    bool doing_precision_landing = precland.target_acquired() && navigating;
+    Vector2f target;
+    plane.precland.get_target_position_relative_cm(target);
+#else
+    bool doing_precision_landing = false;
+#endif
+
+    if ((poscontrol.state == QPOS_POSITION2 && plane.auto_state.wp_distance < 2) ||
+        (doing_precision_landing && target.length() < 2)) {
```
Comment (lukedempsey) [inline ArduPlane/quadplane.cpp]: There is similar code in control_loiter (lines 1231-1257) which was previously written that also controls landing state transitions. Wasn't sure to put it there or here - where there is a QPOS_LAND_DESCEND transition already defined.

File: ArduPlane/quadplane.cpp [hunk 2]
```diff
@@ -2177,6 +2243,23 @@ void QuadPlane::vtol_position_controller(void)
         } else {
             pos_control->set_xy_target(poscontrol.target.x, poscontrol.target.y);
         }
+
+#if PRECISION_LANDING == ENABLED
+        // run precision landing
```
Comment (lukedempsey) [inline ArduPlane/quadplane.cpp]: I'm not fully around pos_control, but it seems that when I override the it with ```pos_control->set_xy_target(target_pos.x, target_pos.y);``` there is less authority to move the craft to the target location - it can hover for many seconds approx 3m away from target coordinates horizontally, delaying the ```Q_POS_DESCEND``` transition

File: libraries/SITL/SIM_QuadPlane.cpp [hunk 3]
```diff
@@ -126,6 +126,7 @@ void QuadPlane::update(const struct sitl_input &input)
     accel_body += quad_accel_body;
 
     update_dynamics(rot_accel);
+    update_external_payload(input);
```
Comment (lukedempsey) [inline libraries/SITL/SIM_QuadPlane.cpp]: I added this so I can get precland working in SITL, it also brings over other SITL instances - I haven't tested them all yet. Is there a reason this wasn't included before?

**Comment by tridge** [comment]
thanks for this work!
It does need a rebase, can you do that or do you want me to do that for you?

**Comment by lukedempsey** [comment]
> thanks for this work!
> It does need a rebase, can you do that or do you want me to do that for you?

No problem, I'll do this tomorrow

File: ArduPlane/Log.cpp [hunk 4]
```diff
@@ -163,6 +163,61 @@ struct PACKED log_Status {
...
+// Write a precision landing entry
+void Plane::Log_Write_Precland()
+{
+ #if PRECISION_LANDING == ENABLED
+    // exit immediately if not enabled
+    if (!precland.enabled()) {
+        return;
+    }
+
+    Vector3f target_pos_meas = Vector3f(0.0f,0.0f,0.0f);
```
Comment (tridge) [inline ArduPlane/Log.cpp]: no need to init these to zero (they auto-init)

File: ArduPlane/Parameters.h [hunk 5]
```diff
@@ -343,6 +343,7 @@ class Parameters {
         k_param_dspoiler_rud_rate,
 
         k_param_logger = 253, // Logging Group
+        k_param_precland = 256,
```
Comment (tridge) [inline ArduPlane/Parameters.h]: I'd rather see this in g2 instead of going over 255

File: ArduPlane/quadplane.cpp [hunk 6]
```diff
@@ -1200,16 +1227,24 @@ void QuadPlane::control_loiter()
...
-            // cut IC engine if enabled
-            if (land_icengine_cut != 0) {
-                plane.g2.ice_control.engine_control(0, 0, 0);
+        float height_above_ground = plane.relative_ground_altitude(plane.g.rangefinder_landing);
+        if (doing_precision_landing) {
+            float max_descent_speed = 0.5f*abs(land_speed_cms);
+            float land_slowdown = MAX(0.0f, pos_control->get_horizontal_error()*(max_descent_speed/precland_acceptable_error));
+            cmb_rate = MIN(-precland_min_descent_speed, -max_descent_speed+land_slowdown);
+            pos_control->set_alt_target_from_climb_rate_ff(cmb_rate, plane.G_Dt, true);
+            } else {
```
Comment (tridge) [inline ArduPlane/quadplane.cpp]: indenting bad, hard to read

File: ArduPlane/quadplane.cpp [hunk 7]
```diff
@@ -1200,16 +1227,24 @@ void QuadPlane::control_loiter()
                                                                   get_desired_yaw_rate_cds());
 
     if (plane.control_mode == &plane.mode_qland) {
-        if (poscontrol.state < QPOS_LAND_FINAL && check_land_final()) {
-            poscontrol.state = QPOS_LAND_FINAL;
-            // cut IC engine if enabled
-            if (land_icengine_cut != 0) {
-                plane.g2.ice_control.engine_control(0, 0, 0);
+        float height_above_ground = plane.relative_ground_altitude(plane.g.rangefinder_landing);
+        if (doing_precision_landing) {
```
Comment (tridge) [inline ArduPlane/quadplane.cpp]: why the different descent rate handling for precland to normal land?
also, it omits the ICE control in precland case

File: ArduPlane/quadplane.cpp [hunk 8]
```diff
@@ -2228,16 +2310,16 @@ void QuadPlane::vtol_position_controller(void)
     }
 
     case QPOS_LAND_DESCEND: {
-        float height_above_ground = plane.relative_ground_altitude(plane.g.rangefinder_landing);
-        pos_control->set_alt_target_from_climb_rate(-landing_descent_rate_cms(height_above_ground),
-                                                    plane.G_Dt, true);
+        cmb_rate = compute_descent_rate(2.0f, doing_precision_landing);
+        pos_control->set_alt_target_from_climb_rate_ff(cmb_rate, plane.G_Dt, true);
```
Comment (tridge) [inline ArduPlane/quadplane.cpp]: why the change to use the _ff method?
can't we just change landing_descent_rate_cms() if we do need a different rate for precland? (and do we?)

File: ArduPlane/quadplane.cpp [hunk 9]
```diff
@@ -2228,16 +2310,16 @@ void QuadPlane::vtol_position_controller(void)
...
         break;
     }
 
-    case QPOS_LAND_FINAL:
-        pos_control->set_alt_target_from_climb_rate(-land_speed_cms, plane.G_Dt, true);
+    case QPOS_LAND_FINAL: {
+        cmb_rate = compute_descent_rate(0.5f, doing_precision_landing);
+        pos_control->set_alt_target_from_climb_rate_ff(cmb_rate, plane.G_Dt, true);
```
Comment (tridge) [inline ArduPlane/quadplane.cpp]: again, why change to _ff, and why change rate at all?

File: ArduPlane/quadplane.cpp [hunk 10]
```diff
@@ -2577,8 +2659,20 @@ bool QuadPlane::verify_vtol_land(void)
...
+    const bool navigating = pos_control->is_active_xy();
+    bool doing_precision_landing = precland.target_acquired() && navigating;
+    Vector2f target;
+    plane.precland.get_target_position_relative_cm(target);
+#else
+    bool doing_precision_landing = false;
+#endif
+
+    if (poscontrol.state == QPOS_POSITION2 && 
+        ((plane.auto_state.wp_distance < 2) || (doing_precision_landing && target.length() < 2))) {
```
Comment (tridge) [inline ArduPlane/quadplane.cpp]: I think we could abstract this into a function that gets the XY distance to target

File: ArduPlane/sensors.cpp [hunk 11]
```diff
@@ -38,6 +38,12 @@ void Plane::read_rangefinder(void)
     rangefinder_height_update();
 }
 
+// return true if rangefinder_alt can be used
+bool Plane::rangefinder_alt_ok(void)
+{
+    return (rangefinder.num_sensors());
```
Comment (tridge) [inline ArduPlane/sensors.cpp]: add a != 0 to change to a bool

File: ArduPlane/precision_landing.cpp [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+{
+    plane.precland.init(scheduler.get_loop_rate_hz());
+}
+
+void Plane::update_precland()
+{
+    int32_t height_above_ground_cm = current_loc.alt;
+
+    // use range finder altitude if it is valid, else try to get terrain alt
+    if (rangefinder_alt_ok()) {
```
Comment (tridge) [inline ArduPlane/precision_landing.cpp]: why not use the existing Plane::relative_ground_altitude() function?

**Comment by lukedempsey** [comment]
> > thanks for this work!
> > It does need a rebase, can you do that or do you want me to do that for you?
> 
> No problem, I'll do this tomorrow

@tridge this branch has previously been kept up to date but merging master into it. Not sure how to squash/rebase this easily. Any advice on what to do? I don't have much experience rebasing.

I'll work through the rest of your notes in the meantime

File: ArduPlane/Parameters.h (hunk 5 above)
Comment (lukedempsey) [reply ArduPlane/Parameters.h]: Is there documentation/example anywhere on how to do the g2 list? Not sure how to implement this. Doesn't look as simple I thought it would be

File: ArduPlane/quadplane.cpp (hunk 8 above)
Comment (lukedempsey) [reply ArduPlane/quadplane.cpp]: Not sure the motivation behind this change, left it in because it seemed to be implicitly approved by nobody complaining about it in the previous PR it appeared in. It doesn't make sense to me, I'll remove it because it doesn't seem in the scope of the PR anyway

File: libraries/SITL/SIM_QuadPlane.cpp (hunk 3 above)
Comment (tridge) [reply libraries/SITL/SIM_QuadPlane.cpp]: that's fine

**Comment by jo-m** [comment]
@lukedempsey is this going to be merged soon?

**Comment by davidbuzz** [comment]
I poked tridge during the dev call, and he said it needs a rebase, and he might have a try....

**Comment by CBMUniversal** [comment]
Is this still being worked on?  Is there a timeline for this being merged?

**Comment by jonathan84clark** [comment]
Hello,
I have halted my work on this feature due to feature creep. The changes I made successfully implemented precision landing in quad plane however when I attempted a pull request the ArduPilot team had numerous refactors, optimizations and other improvements they wanted to add before they would allow it to be merged. I was not comfortable implementing the myriad of changes as I no longer have a platform I can test my changes on.
Thanks,
Jonathan L Clark 

    On Monday, October 28, 2019, 11:50:40 AM PDT, CBMUniversal <notifications@github.com> wrote:  
 
 
Is this still being worked on? Is there a timeline for this being merged?

**Comment by rmackay9** [comment]
I'm not responsible for the plane code but in general a good way to escalate a PR (give it a push towards getting merged) is to add the "DevCallTopic" label (or ask someone like me, tridge, amilcarlucas, etc to add it) and then attend the weekly dev call.  I know the timing of the weekly call can be difficult for some locations... but just trying to give some help on the process.

mumble server used for weekly dev call: http://ardupilot.org/dev/docs/ardupilot-mumble-server.html

**Comment by jonathan84clark** [comment]
I should be clear; I have no intention of completing all of the requested changes and features that were recommended when I submitted this pull request. I personally feel that those requests are outside the scope of my task; implementing precision landing in quadplane and represent scope creep.
Additionally it would be extremely high-risk for me to make any additional changes as I no longer have a platform to test on. 
I stand by my original code change; it worked on the day I submitted my pull request. I have documented video evidence of its functionality. However without access to test platforms I cannot confirm it will work today (I can't deny it either).

**Comment by tridge** [comment]
I have rebased this on master

**Comment by peterbarker** [comment]
I've rebased this on master, having already merged some of the patches (which needed @rmackay9 's approval)

I've also fixed several things I spotted - the update rate we pass into AC_PrecLand on construction must be the same as that in the scheduler table; the library itself takes the minimum of that and the scheduler loop rate to determine its buffer sizes.

There are still several more issues to cross here.

File: ArduPlane/precision_landing.cpp (hunk 12 above)
Comment (peterbarker) [reply ArduPlane/precision_landing.cpp]: I'm quite sure we shouldn't do this - and should stick to passing rangefinder data through to the library here.

If we were to drop down to using the terrain database the errors could be really quite large. Better to be safe here and solely rely on rangefinder data IMO - at least for an initial cut.

**Comment by ddomit** [comment]
> I've rebased this on master, having already merged some of the patches (which needed @rmackay9 's approval)
> 
> I've also fixed several things I spotted - the update rate we pass into AC_PrecLand on construction must be the same as that in the scheduler table; the library itself takes the minimum of that and the scheduler loop rate to determine its buffer sizes.
> 
> There are still several more issues to cross here.

Hi peter! Just wondering if there is anything i can help with to have this feature implemented into quadplane

**Comment by davidbuzz** [comment]
i think if @tridge  and @peterbarker  have both massaged the code thru forced-pushes AND tridge has written 'looks good' , then this is simply a merge-on-ci-pass now.

**Comment by peterbarker** [comment]
I'm removing MergeOnCIPass.  At the very least the scope of rebasing required would mean this would need another thorough review.

**Comment by markcfong561** [comment]
Is this going to be merge to master anytime soon?

**Comment by julled** [comment]
+1 i would be also very interested in getting this into mainline

**Comment by tridge** [comment]
i've decided to do this with lua scripts for flexibility: https://github.com/ArduPilot/ardupilot/pull/26380
